pkg/container/versioned/value_test.go

```diff
@@ -0,0 +1,388 @@
+	v = value1.At(handle3)
+	assert.Equal(t, []uint32{100, 150, 200}, v)
+
+	v = value2.At(handle3)
+	assert.Equal(t, []uint32{110, 150, 190}, v)
+
+	// Old handle sees the previous values
+	v = value1.At(handle2)
+	assert.Equal(t, []uint32{100, 200}, v)
+
+	v = value2.At(handle2)
+	assert.Equal(t, []uint32{110, 190}, v)
+
+	// first handle still sees no value
+	assert.Empty(t, value1.At(handle))
+
+	// delete the value1 at next version
+	tx = cv.PrepareNextVersion()
+	assert.Nil(t, value1.RemoveAt(tx))
+	assert.Nil(t, tx.Commit())
+
+	// new handle sees an empty value1, but value2 remains
+	handle4 := cv.GetVersionHandle()
+	assert.Empty(t, value1.At(handle4))
+
+	v = value2.At(handle4)
+	assert.Equal(t, []uint32{110, 150, 190}, v)
+
+	assert.Empty(t, value1.At(handle))
+	assert.Empty(t, value2.At(handle))
+
+	v = value1.At(handle3)
+	assert.Equal(t, []uint32{100, 150, 200}, v)
+
+	v = value2.At(handle3)
+	assert.Equal(t, []uint32{110, 150, 190}, v)
+
+	v = value1.At(handle2)
+	assert.Equal(t, []uint32{100, 200}, v)
+
+	v = value2.At(handle2)
+	assert.Equal(t, []uint32{110, 190}, v)
+
+	// handle closers can be called in any order
+	assert.Nil(t, handle2.Close())
+	assert.Nil(t, handle3.Close())
+	assert.Nil(t, handle1.Close())
+	assert.Nil(t, handle4.Close())
+
+	// old value1 have been cleaned off
+	cleaner.waitUntilCleaned(handle3.version)
+	assert.Nil(t, value1.head.Load())
+
+	// but value2 remains, as it was not removed
+	assert.NotNil(t, value2.head.Load())
+}
+
+func TestPairSlice(t *testing.T) {
+	cv := Coordinator{}
+	s := make(VersionedSlice[int], 0, 10)
+
+	// 1st value '2000' at version 1
+	tx := cv.PrepareNextVersion()
+	s = s.Append(2000, tx)
+	assert.Len(t, s, 1)
+	assert.Equal(t, version(1), s[0].version)
+	assert.Equal(t, 2000, s[0].value)
+
+	tx.Commit()
+	tx = cv.PrepareNextVersion()
+
+	// 2nd value '1000' at version 2
+	s = s.Append(1000, tx)
+	assert.Len(t, s, 2)
+	assert.Equal(t, version(1), s[0].version)
+	assert.Equal(t, 2000, s[0].value)
+	assert.Equal(t, version(2), s[1].version)
+	assert.Equal(t, 1000, s[1].value)
+
+	tx.Commit()
+	tx = cv.PrepareNextVersion()
+
+	// 3rd value '3000' at version 3
+	s = s.Append(3000, tx)
+	assert.Len(t, s, 3)
+	assert.Equal(t, version(1), s[0].version)
+	assert.Equal(t, 2000, s[0].value)
+	assert.Equal(t, version(2), s[1].version)
+	assert.Equal(t, 1000, s[1].value)
+	assert.Equal(t, version(3), s[2].version)
+	assert.Equal(t, 3000, s[2].value)
+
+	// get all values before version 3; excluding the last value
+	var values []int
+	n := s.ForEachBefore(KeepVersion(3), func(i int) {
+		values = append(values, i)
+	})
+	assert.Equal(t, []int{2000, 1000}, values)
+	s = s[n:]
+
+	// get all values upto and including maxVersion; get the lone value
+	values = nil
+	n = s.ForEachBefore(KeepVersion(invalidVersion), func(i int) {
+		values = append(values, i)
+	})
+	assert.Equal(t, []int{3000}, values)
+	s = s[n:]
+	assert.Len(t, s, 0)
+}
+
+func TestVersionedChaos(t *testing.T) {
```

Comment:
The race detector is pretty unhappy with this test - `go test -c -race ./pkg/container/versioned && ./versioned.test` gives me immediate hits of the form:

```
==================
WARNING: DATA RACE
Read at 0x00c000074488 by goroutine 148:
  github.com/cilium/cilium/pkg/container/versioned.(*Coordinator).GetVersionHandle()
      /home/bimmlerd/wrk/cilium/pkg/container/versioned/value.go:342 +0x84
  github.com/cilium/cilium/pkg/container/versioned.TestVersionedChaos.func2()
      /home/bimmlerd/wrk/cilium/pkg/container/versioned/value_test.go:337 +0x2c

Previous write at 0x00c000074488 by goroutine 114:
  github.com/cilium/cilium/pkg/container/versioned.(*Coordinator).commit()
      /home/bimmlerd/wrk/cilium/pkg/container/versioned/value.go:242 +0xac
  github.com/cilium/cilium/pkg/container/versioned.(*Tx).Commit()
      /home/bimmlerd/wrk/cilium/pkg/container/versioned/value.go:211 +0x3d0
  github.com/cilium/cilium/pkg/container/versioned.TestVersionedChaos.func3()
      /home/bimmlerd/wrk/cilium/pkg/container/versioned/value_test.go:367 +0x3ac

Goroutine 148 (running) created at:
  github.com/cilium/cilium/pkg/container/versioned.TestVersionedChaos()
      /home/bimmlerd/wrk/cilium/pkg/container/versioned/value_test.go:336 +0x730
  testing.tRunner()
      /usr/lib/go-1.22/src/testing/testing.go:1689 +0x180
  testing.(*T).Run.gowrap1()
      /usr/lib/go-1.22/src/testing/testing.go:1742 +0x40

Goroutine 114 (finished) created at:
  github.com/cilium/cilium/pkg/container/versioned.TestVersionedChaos()
      /home/bimmlerd/wrk/cilium/pkg/container/versioned/value_test.go:344 +0x384
  testing.tRunner()
      /usr/lib/go-1.22/src/testing/testing.go:1689 +0x180
  testing.(*T).Run.gowrap1()
      /usr/lib/go-1.22/src/testing/testing.go:1742 +0x40
==================
```

Reply:
A `defer` was missing do a lock was released right after closing, rather than at the end of `commit()`, thanks for noticing!